### PR TITLE
[Gst/MQTT] Support dynamic Caps re-negotiation according to the Caps information sent from the other node @open sesame 05/19 13:00

### PR DESCRIPTION
**Pull request by wooksong**
The purpose of this PR is to support just-in-time capability negotiation in the GStreamer pipeline using the capability information included in the message to which the other pipeline node sent.

In detail, this PR contains the following changes:
For the mqttsink, 
 - The message header is modified to include a string representation of the fixated GstCaps.
 For the mqttsrc,
 - A procedure extracting the GstCaps information from the incoming header is added.
 - According to the extracted GstCaps information, the pipeline tries to negotiation the capabilities just before processing the first buffer.

Signed-off-by: Wook Song <wook16.song@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/mqtt/mqttsrc.c [hunk 1]
```diff
@@ -485,6 +489,67 @@ gst_mqtt_src_get_caps (GstBaseSrc * basesrc, GstCaps * filter)
...
+
+      if (fixed_caps && gst_caps_is_fixed (fixed_caps)) {
+        result = gst_base_src_set_caps (basesrc, fixed_caps);
+        if (peercaps == self->caps)
+          gst_caps_unref (fixed_caps);
+      }
+    }
+    gst_caps_unref (caps);
+  } else {
+    result = FALSE;
```
Comment (dongju-chae) [inline gst/mqtt/mqttsrc.c:537]: Q. no need to unref caps when `gst_caps_is_empty()`?

File: gst/mqtt/mqttcommon.h [hunk 2]
```diff
@@ -21,8 +21,8 @@
 #define GST_MQTT_ELEM_NAME_SINK "mqttsink"
 #define GST_MQTT_ELEM_NAME_SRC "mqttsrc"
 
-#define GST_MQTT_LEN_MSG_HDR    512
-
+#define GST_MQTT_LEN_MSG_HDR          1024
+#define GST_MQTT_MAX_LEN_GST_CPAS_STR 512
```
Comment (myungjoo) [inline gst/mqtt/mqttcommon.h:25]: Are you going to make this dynamic later? (non-fixed size of header)

File: gst/mqtt/mqttsink.c [hunk 3]
```diff
@@ -764,6 +772,27 @@ gst_mqtt_sink_event (GstBaseSink * basesink, GstEvent * event)
...
+{
+  GstMqttSink *self = GST_MQTT_SINK (basesink);
+  gboolean ret;
+
+  ret = gst_caps_replace (&self->in_caps, caps);
+
+  if (ret && gst_caps_is_fixed (self->in_caps)) {
+    char *caps_str = gst_caps_to_string (caps);
+
+    strncpy (self->mqtt_msg_hdr.gst_caps_str, caps_str, strlen (caps_str));
```
Comment (myungjoo) [inline gst/mqtt/mqttsink.c:789]: min(strlen(caps_str), GST_MQTT_MAX_LEN_GST_CPAS_STR-1) ?

**Review by myungjoo — APPROVED** [review]

File: gst/mqtt/mqttcommon.h (hunk 2 above)
Comment (wooksong) [reply gst/mqtt/mqttcommon.h:25]: > Are you going to make this dynamic later? (non-fixed size of header)

Yes, I am.

File: gst/mqtt/mqttsrc.c (hunk 1 above)
Comment (wooksong) [reply gst/mqtt/mqttsrc.c:537]: It is required. Thanks.

File: gst/mqtt/mqttsink.c (hunk 3 above)
Comment (wooksong) [reply gst/mqtt/mqttsink.c:789]: That would be better. Thanks.

**Review by helloahn — APPROVED** [review]
LGTM

**Review by dongju-chae — APPROVED** [review]
LGTM

**Pull request merged**